### PR TITLE
[JavaScript] Implement new Concept Exercise: concept/array-analysis #1370

### PR DESCRIPTION
**Pull request by peterchu999**

### CODE REVIEW

**Comment by SleeplessByte** [comment]
Can you rebase this branch on `master` so those previous commits disappear?

File: languages/javascript/exercises/concept/arrays-analysis/.docs/introduction.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+In Javascript, an array is a list-like structure with no fixed length which can hold any type of primitives or ojects, even mixed types. The array elements can be accesed by their index, array also provide a bunch of built-in methods. in this exercise we will tackle the build-in method to analyse the array
+
+Example of analysis without looping through it:
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays-analysis/.docs/introduction.md:3]: Let's give the example using a `for` loop and _then_ the example using `indexOf`.
Index of still _loops_, but it's easier to _comprehend_ --

**Comment by peterchu999** [comment]
okay @SleeplessByte i'll try to do it as fast as i can.

**Comment by peterchu999** [comment]
@SleeplessByte so sorry if i make any mistake when i'm rebasing. i've add as your suggestion

File: languages/javascript/exercises/concept/arrays-analysis/.meta/config.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "contributors": [
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/arrays-analysis/.meta/config.json:2]: > Also, you're missing a .meta/config.json file with you as author and contributor :)

@SleeplessByte We didn't specify this yet, but our thought was to aim for no duplication here. So if you're an author and a contributor, you're just listed as an author.

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/arrays-analysis/.meta/config.json:2]: Oh; at some point the exact opposite was said. I obvi don't care as long as credit is given :)

**Comment by SleeplessByte** [comment]
@ErikSchierboom recommendation on naming?

`arrays-analysis` vs `array-analysis` vs `arrays-x`?

File: languages/javascript/exercises/concept/arrays-analysis/.meta/config.json (hunk 2 above)
Comment (ErikSchierboom) [reply languages/javascript/exercises/concept/arrays-analysis/.meta/config.json:2]: Oh, we'll have to figure this out. Ignore it for now, we can always fix it later if the other option was indeed the correct one :)

CC @iHiD

**Comment by ErikSchierboom** [comment]
Any reasons not to just name this `arrays`? :)

**Comment by peterchu999** [comment]
@ErikSchierboom we can't name it `arrays` because the previous exercise on array already take that name

**Comment by ErikSchierboom** [comment]
Ah, right. I think `array-analysis` is fine then. We can always rename later.

**Comment by SleeplessByte** [comment]
Let use `array-analysis`, given that it's referring to the Concept of Array analysis. @peterchu999 can you rename?

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  throw new Error('Implement the getCardIndex function')
+}
+
+/**
+ * Check if the array contains cardGiven
+ *
+ * @param {number[]} array
+ * @param {number} cardGiven
+ *
+ * @returns {boolean} Whether contains a cardGiven
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:19]:
```suggestion
 * @returns {boolean} true if card is in the stack, false otherwise
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function isStackIncludes(array, cardGiven) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:21]:
```suggestion
export function doesStackIncludesCard(stack, card) {
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  throw new Error('Implement the isStackIncludes function')
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:22]:
```suggestion
  throw new Error('Implement the doesStackIncludesCard function')
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * Check if every value in array was even
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:26]:
```suggestion
 * Determine if each card is even
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @param {number[]} array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:28]:
```suggestion
 * @param {number[]} stack
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @returns {boolean} Whether every value in array was even
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:30]:
```suggestion
 * @returns {boolean} true if each value is even, false otherwise
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function isEvenStack(array) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:32]:
```suggestion
export function isEachCardEven(stack) {
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  throw new Error('Implement the isEvenStack function')
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:33]:
```suggestion
  throw new Error('Implement the isEachCardEven function')
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * Check if at least one odd value card
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:37]:
```suggestion
 * Check if the stack contains an odd card
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @param {number[]} array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:39]:
```suggestion
 * @param {number[]} stack
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @returns {boolean} Whether array contains odd card
+ */
+export function isStackContainsOdd(array) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:43]:
```suggestion
export function doesStackIncludesOddCard(stack) {
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+  throw new Error('Implement the isStackContainsOdd function')
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:44]:
```suggestion
  throw new Error('Implement the doesStackIncludesOddCard function')
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * Get the first odd value card from array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:48]:
```suggestion
 * Get the first odd card from the stack
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @param {number[]} array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:50]:
```suggestion
 * @param {number[]} stack
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @returns {number} First odd value card
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:52]:
```suggestion
 * @returns {number} the first odd value
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function getFirstOddCard(array) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:54]:
```suggestion
export function getFirstOddCard(stack) {
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * Get the index of first even value card from array
+ *
+ * @param {number[]} array
+ *
+ * @returns {number} First even value card index
+ */
+export function getFirstEvenCardIndex(array) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:65]:
```suggestion
export function getFirstEvenCardPosition(stack) {
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  throw new Error('Implement the getFirstEvenCardIndex function')
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:66]:
```suggestion
  throw new Error('Implement the getFirstEvenCardPosition function')
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+ * @param {number[]} array
+ *
+ * @returns {number} First odd value card
+ */
+export function getFirstOddCard(array) {
+  throw new Error('Implement the getFirstOddCard function')
+}
+
+/**
+ * Get the index of first even value card from array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:59]:
```suggestion
 * Determine the position of the first card that is even
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @param {number[]} array
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:61]:
```suggestion
 * @param {number[]} stack
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @returns {number} First even value card index
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:63]:
```suggestion
 * @returns {number} position of the first card that is even
```

File: languages/javascript/exercises/concept/array-analysis/array-analysis.spec.js [hunk 24]
```diff
@@ -0,0 +1,93 @@
+import {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.spec.js:1]: Please update these method names according to the suggestions to the stub file.

- [ ] imports
- [ ] test names
- [ ] test contents

File: languages/javascript/exercises/concept/array-analysis/example.js [hunk 25]
```diff
@@ -0,0 +1,23 @@
+export function getCardIndex(array, card) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/example.js:1]: Please update this example file according to the changes to the tests file and stub file.

File: languages/javascript/exercises/concept/array-analysis/.meta/config.json [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+      "exercism_username": "peterchu999"
+    }
+  ],
+  "authors": [
+    {
+      "github_username": "peterchu999",
+      "exercism_username": "peterchu999"
+    }
+  ]
+}
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-analysis/.meta/config.json:14]: Can you add me (`SleeplessByte`) to the contributors?

**Comment by peterchu999** [comment]
okay @SleeplessByte will do it as fast as i can

File: languages/javascript/exercises/concept/array-analysis/.meta/config.json (hunk 26 above)
Comment (peterchu999) [reply languages/javascript/exercises/concept/array-analysis/.meta/config.json:14]: sure

**Comment by SleeplessByte** [comment]
No rush! Take your time @peterchu999  :)

File: languages/javascript/config.json [hunk 27]
```diff
@@ -33,6 +33,12 @@
         "concepts": ["arrays"],
         "prerequisites": ["numbers"]
       },
+      {
+        "slug": "array-analysis",
+        "uuid": "45d956db-d4ef-4468-b1d3-47021f172c15",
+        "concepts": ["array-analysis"],
+        "prerequisites": ["arrays", "booleans", "numbers", "callback"]
```
Comment (SleeplessByte) [inline languages/javascript/config.json:40]:
```suggestion
        "prerequisites": ["arrays", "booleans", "numbers", "callbacks"]
```

File: languages/javascript/exercises/concept/array-analysis/.docs/after.md [hunk 28]
```diff
@@ -0,0 +1,18 @@
+In JavaScript, instances of an `Array` are regular objects, decorated with lots of [useful methods][array-docs] from the `Array.prototype`. It includes the `length` property and also for traversing and mutating the array. Its elements are indexed properties of that object.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/.docs/after.md:1]:
```suggestion
In JavaScript, instances of an `Array` are regular objects, decorated with lots of [useful methods][array-docs] from the `Array.prototype`. It includes the `length` property and also methods for traversing and mutating the array. Its elements are indexed properties of that object.
```

File: languages/javascript/exercises/concept/array-analysis/.docs/introduction.md [hunk 29]
```diff
@@ -0,0 +1,23 @@
+In JavaScript, an array is a list-like structure with no fixed length which can hold any type of primitives or objects, or even mixed types. The array elements can be accessed by their index. Arrays are also given a bunch of built-in methods.
+
+In this exercise we will use built-in methods to analyse the contents of an array, instead of using a `for` loop, or `.forEach`.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/.docs/introduction.md:3]: This comment refers to the exercise itself, which means that it should be part of the `instructions.md` file instead of the `introduction.md` file.

File: languages/javascript/exercises/concept/array-analysis/.docs/introduction.md [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+const numbers = [1, 'two', 3, 'four']
+for (var i = 0; i < numbers.length; i++) {
+    if (numbers[i] === 'two'){
+        return numbers.[i]
+    }
+}
+// => 1
+```
+
+Example of analysis using a build-in method:
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/.docs/introduction.md:17]:
```suggestion
Example of analysis using a built-in method:
```

File: languages/javascript/exercises/concept/array-analysis/.docs/introduction.md [hunk 31]
```diff
@@ -0,0 +1,23 @@
+In JavaScript, an array is a list-like structure with no fixed length which can hold any type of primitives or objects, or even mixed types. The array elements can be accessed by their index. Arrays are also given a bunch of built-in methods.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/.docs/introduction.md:1]: I like how you've managed to make this file very concise!

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Determine if the stack contains the card
+ *
+ * @param {number[]} stack
+ * @param {number} card
+ *
+ * @returns {boolean} true if card is in the stack, false otherwise
+ */
+export function doesStackIncludesCard(stack, card) {
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:21]:
```suggestion
export function doesStackIncludeCard(stack, card) {
```

Could you also fix this minor spelling mistake in the other places where this text occurs?

File: languages/javascript/exercises/concept/array-analysis/array-analysis.js [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * Check if stack conatains odd-value card
+ *
+ * @param {number[]} stack
+ *
+ * @returns {boolean} Whether array contains odd card
+ */
+export function doesStackIncludesOddCard(stack) {
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/array-analysis.js:43]:
```suggestion
export function doesStackIncludeOddCard(stack) {
```

File: languages/javascript/exercises/concept/array-analysis/.docs/hints.md [hunk 34]
```diff
@@ -0,0 +1,31 @@
...
+- Array indices starts at `0`.
+- There is a [built-in][indexof_method_docs] method to get given value index.
+
+### 2. Check whether the given card is in the stack
+
+- There is a [built-in][includes_method_docs] method to check whether the value is present in the array
+
+### 3. Check whether every card in stack is even number
+
+- There is a [built-in][every_method_docs] method to check if each value passes a predicate
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-analysis/.docs/hints.md:12]: I'm wondering if students understand what a _predicate_ is. I'm guessing some of them don't. Maybe rephrase a bit, introduce the term in the introduction or link to a document explaining what a predicate is?

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md [hunk 35]
```diff
@@ -0,0 +1,59 @@
+Elyse, magician-to-be, continues her training. She will be given several stacks of cards that she need to perform her tricks on, without looking at each card one by one.
```
Comment (kotp) [inline languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:1]:
```suggestion
Elyse, magician-to-be, continues her training. She will be given several stacks of cards that she needs to perform her tricks.
```

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md [hunk 36]
```diff
@@ -0,0 +1,59 @@
...
+
+To make things a bit easier she only uses the cards 1 to 10.
```
Comment (kotp) [inline languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:3]:
```suggestion
To make things a bit easier, she only uses the cards 1 to 10.
```

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md [hunk 37]
```diff
@@ -0,0 +1,59 @@
...
+
+### 1. Find the position of a card
+
+Elyse needs the position (index) of a card in the stack
```
Comment (kotp) [inline languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:7]:
```suggestion
Elyse wants to know the position (index) of a card in the stack
```

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md [hunk 38]
```diff
@@ -0,0 +1,59 @@
...
+
+```javascript
+const card = 2
+getCardPosition([9, 7, 3, 2], card)
+// => 3
+```
+
+### 2. Check for presence of a card
+
+Elyse needs to determine if a card is present in the stack
```
Comment (kotp) [inline languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:17]:
```suggestion
Elyse wants to determine if a card is present in the stack
```

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+Elyse wants to know if every card is even
+
+```javascript
+isEachCardEven([2, 4, 6, 7])
+// => false
+```
+
+### 4. Check if a stack contains an odd number
+
+Elyse needs to know if there is an odd number in the stack
```
Comment (kotp) [inline languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:36]:
```suggestion
Elyse wamts to know if there is an odd number in the stack
```

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:36]: wamts 🗡️

File: languages/javascript/exercises/concept/array-analysis/.docs/hints.md (hunk 34 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-analysis/.docs/hints.md:12]: I don't think we should rephrase. I do think we _could_ explain it again in `introduction.md`, BUT this exercise relies on `callbacks` as a Concept, which should explain this.

File: languages/javascript/exercises/concept/array-analysis/.docs/instructions.md (hunk 39 above)
Comment (kotp) [reply languages/javascript/exercises/concept/array-analysis/.docs/instructions.md:36]: Wat U tokkin bout? ;)

File: languages/javascript/exercises/concept/array-analysis/.docs/hints.md (hunk 34 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-analysis/.docs/hints.md:12]: (I'll just write a small blog post on this and put it up -- how's that?)

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-analysis/.docs/hints.md:12]: https://derk-jan.com/2020/05/predicate/

**Comment by peterchu999** [comment]
wow lot of review , then i'll use @SleeplessByte article link in introductions.md

**Comment by ErikSchierboom** [comment]
@peterchu999 I think they're all (very) minor things, which you should absolutely take as a compliment!

File: languages/javascript/exercises/concept/array-analysis/.docs/introduction.md (hunk 29 above)
Comment (peterchu999) [reply languages/javascript/exercises/concept/array-analysis/.docs/introduction.md:3]: which part should i move ? is it starting from "In this exercise we will ...." or the whole instruction.md ?

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-analysis/.docs/introduction.md:3]: > In this exercise we will use built-in methods to analyse the contents of an array, instead of using a `for` loop, or `.forEach`.

**Comment by aimorris** [comment]
I've marked all the review comments that you've addressed as resolved. Can you fix up the last few bits?

**Comment by peterchu999** [comment]
Hi @aimorris i think i have fix the issue according to review, and what do you mean fix lat few thing? can you make a review so that i can know what is wrong with my code

**Comment by ErikSchierboom** [comment]
@peterchu999 I think @aimorris is referring to some open comments:

- https://github.com/exercism/v3/pull/1388/files#r420025870
- https://github.com/exercism/v3/pull/1388/files#r420588120
- https://github.com/exercism/v3/pull/1388/files#r420029031

That said, I don't think they are necessarily deal-breakers. We could also merge this PR and then fix any of those issues later. @SleeplessByte what do you think?

**Comment by SleeplessByte** [comment]
Yeah sure; I'll quickly update the references offline.
